Add policy enforcer to sanity check on policy in query execution

cecemei opened the pull request:
### Description
Added `PolicyEnforcer` interface, `NoopPolicyEnforcer` and `RestrictAllTablesPolicyEnforcer`. It'd be configurable by `druid.policy.enforcer`, through `PolicyModule`.

- On query planning/execution on broker, `PolicyEnforcer` works in building `DruidQuery` and authorization step in `QueryLifecycle`, `withPolicies(Map<String, Optional<Policy>> policyMap, PolicyEnforcer policyEnforcer)`. Specifically, `enforcer.validateOrElseThrow(TableDataSource ds, Policy policy)` throws an exception if validation fails.
- On query runner/processor, `PolicyEnforcer` is passed in `ExecutionVertex.createSegmentMapFunction(PolicyEnforcer)`, when the mapFn is called, it validates the mapped segment called from `SegmentReference.ValidateOrElseThrow(PolicyEnforcer policyEnforcer)`. Specifically,
  - for `HashJoinSegment` and `WrappedSegment`, call delegate.validateOrElseThrow
  - for `RestrictedSegment` and `ReferenceCountingSegment`, call enforcer to validate with a `ReferenceCountingSegment` and policy (null for ReferenceCountingSegment not wrapped in RestrictedSegment.

As a singleton object, the enforcer is injected to `QueryLifecycleFactory` and `TaskToolboxFactory`, and made its way down to `SinkQuerySegmentWalker`, `FrameContext` and `QueryLifecycle`. 

Additionally,
- refactored `ServerManagerTest` to use Guice bindings (new test dependency), and `TestSegmentCacheManager` for loading segments.
- added a `MSQTaskQueryMakerTest` class for easy binding of enforcer. 


<hr>

##### Key changed/added classes in this PR
 * `PolicyEnforcer`
 * `NoopPolicyEnforcer` (with test class too)
 * `RestrictAllTablesPolicyEnforcer` (with test class too)
 * `PolicyModule` (with test class too)
 * `ServerManagerTest`
 * `MSQTaskQueryMakerTest`

<hr>


This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

kgyrtkirk commented on processing/src/main/java/org/apache/druid/query/policy/PolicyConfig.java:
```diff
@@ -0,0 +1,144 @@
+  /**
+   * Defines how strict we want to enforce the policy on tables during query execution process.
+   * <ol>
+   *   <li>{@code APPLY_WHEN_APPLICABLE}, the most basic level, restriction is applied whenever seen fit.
```
I think this line should be an apidoc on the enum value

kgyrtkirk commented on processing/src/main/java/org/apache/druid/query/policy/PolicyConfig.java:
```diff
@@ -0,0 +1,144 @@
+   */
+  public enum TablePolicySecurityLevel
+  {
+    @JsonProperty("0") APPLY_WHEN_APPLICABLE(0f),
```
I believe this name is a bit unfortunate; I feel like this wants to be something like "allow all when not blocked"
if yes then it could be: `ALLOW_MISSING_POLICY`

kgyrtkirk commented on processing/src/main/java/org/apache/druid/query/policy/PolicyConfig.java:
```diff
@@ -0,0 +1,144 @@
+   * Defines how strict we want to enforce the policy on tables during query execution process.
+   * <ol>
+   *   <li>{@code APPLY_WHEN_APPLICABLE}, the most basic level, restriction is applied whenever seen fit.
+   *   <li>{@code POLICY_CHECKED_ON_ALL_TABLES_POLICY_MUST_EXIST}, every table must have a policy when requests come from external users.
```
I think a shorter one would be: `MANDATORY` ; or `ENFORCED`

kgyrtkirk commented on processing/src/main/java/org/apache/druid/query/policy/PolicyConfig.java:
```diff
@@ -0,0 +1,144 @@
+   * Returns true if the security level requires that, every table must have a policy during query execution stage,
+   * this means the table must have a non-empty value in the policy map.
+   */
+  public boolean policyMustBeCheckedAndExistOnAllTables()
```
I was wondering about an alternate way of implementing this:
* `PolicyConfig` is a dispatcher
* it has a list of `PolicyValidator` -s ; maybe those classes are instantiated with the default constructor?
* the `allowPolicy` and other methods are dispatched to the ones below
* `PolicyConfig` could also implement the `PolicyValidatory` interface

I was also wondering if the whole `PolicyConfig` is a too much - wouldn't it be enough to enable the users to just bind a `PolicyValidator`  they want to use; by specifying its classname somehow - and leave the dispatch or not / etc problems alone :)

gianm replied:
IMO, "enable users [operators, really] to bind a PolicyValidator they want to use" is the right way to go. Although I would suggest naming it `PolicyEnforcer`. It should be specified the same way most potentially-complex choices are specified: using a Jackson-deserializable interface bound with `JsonConfigProvider`. That makes it possible to provide custom enforcers in site-specific extensions, and those custom enforcers can themselves take configuration.

I suggest the `PolicyEnforcer` interface should have two methods:

- `validateQuery(Query)` which runs on both Brokers and data servers before a query is issued.
- `validateProcessableSegment(Segment)` which runs on data servers on the mapped segments that result from `createSegmentMapFunction`.

We should provide some basic builtin validator in core, mostly so we can write proper tests for the validation feature. IMO it would work to provide two builtin enforcers:

- `none` (default enforcer, does nothing)
- `restrictAllTables` (enforcers that all tables are wrapped in some `restricted` datasource)

With the `JsonConfigProvider` approach, these could be specified at runtime using something like `druid.auth.policy.enforcer.type = restrictAllTables`.

cecemei replied:
Updated to use a `PolicyEnforcer` with two subtypes `NoopPolicyEnforcer` and `RestrictAllTablesPolicyEnforcer`, and also a `PolicyModule` to bind a singleton instance from `Properties`.

kgyrtkirk commented:
left a few comments; happy to talk about it further!

cecemei commented on server/src/test/java/org/apache/druid/server/coordination/ServerManagerTest.java:
```diff
@@ -464,7 +430,49 @@ public void testMultipleDrops() throws Exception
-  public void testGetQueryRunnerForIntervalsWhenTimelineIsMissingReturningNoopQueryRunner()
+  public void testReferenceCounting_restrictedSegment() throws Exception
+  {
+    factory = new MyQueryRunnerFactory(new CountDownLatch(1), new CountDownLatch(1), new CountDownLatch(2));
+    conglomerate = DefaultQueryRunnerFactoryConglomerate.buildFromQueryRunnerFactories(ImmutableMap.of(
+        SearchQuery.class,
+        factory
+    ));
+    serverManager = Guice.createInjector(BoundFieldModule.of(this)).getInstance(ServerManager.class);
```
yea the notifyLatch in factory was changed to 2, so we can get the expected result for `factory.getSegmentReferences().get(0).getNumReferences())`

cecemei commented on server/src/test/java/org/apache/druid/server/coordination/ServerManagerTest.java:
```diff
@@ -123,103 +118,77 @@
+    serverConfig = new ServerConfig();
+    authConfig = new AuthConfig();
+
+    Guice.createInjector(BoundFieldModule.of(this)).injectMembers(this);
```
`BoundFieldModule` just creates a module from all the Bind annotation in the current class, the same thing if we do `binder -> binder.bind().toInstance()`. It does introduce another dependency guice:test-lib. Do you have any concerns on this type of injecting?

cecemei commented on server/src/test/java/org/apache/druid/server/coordination/SegmentBootstrapperTest.java:
```diff
@@ -37,6 +37,7 @@
 import org.apache.druid.segment.loading.StorageLocationConfig;
 import org.apache.druid.server.SegmentManager;
 import org.apache.druid.server.metrics.DataSourceTaskIdHolder;
+import org.apache.druid.test.utils.TestSegmentCacheManager;
```
removed the style changes

cecemei commented on processing/src/main/java/org/apache/druid/query/policy/PolicyConfig.java:
```diff
@@ -0,0 +1,144 @@
+
+  public boolean allowPolicy(@Nonnull Policy policy)
+  {
+    return allowedPolicies.isEmpty() || allowedPolicies.contains(policy.getClass().getSimpleName());
```
updated to use reflections to validate class is subtype of `Policy`

cecemei commented on server/src/main/java/org/apache/druid/server/coordination/ServerManager.java:
```diff
@@ -190,6 +195,11 @@ public <T> QueryRunner<T> getQueryRunnerForSegments(Query<T> theQuery, Iterable<
+    if (!(theQuery instanceof SegmentMetadataQuery)
+        && !dataSourceFromQuery.validate(authConfig.getTableSecurityPolicyConfig())) {
+      throw new ISE("Failed security validation with dataSource [%s]", dataSourceFromQuery);
+    }
```
i think in my mind this special case for `SegmentMetadataQuery` is a temporary solution, we're likely going to figure out how RestrictedDS work on that later. we also have this special case handling for `QueryLifeCycle` as well.

cecemei replied:
removed special handling for it, as long as https://github.com/apache/druid/pull/17864 is merged in before this one

cecemei commented:
> left a few comments; happy to talk about it further!

Addressed your comments, please take a look again, thx!

cryptoe commented:
@cecemei it might be better to do a merge instead of a rebase. That way all the commits will not show up in commit history.

gianm commented on processing/src/main/java/org/apache/druid/query/DataSource.java:
```diff
@@ -140,6 +141,16 @@ default DataSource withPolicies(Map<String, Optional<Policy>> policyMap)
+  /**
+   * Returns true if the datasource complies with the policy restrictions on tables.
+   */
+  default boolean validate(PolicyEnforcer policyEnforcer)
```
It looks like this method is unused. We won't need it anyway, since the enforcer is able to inspect the query's datasource directly. Please remove it.

cecemei replied:
actually updated walkers to use query.datasource().validate(enforcer). this should handle UnionDS UnnestDS as well.

gianm commented on processing/src/main/java/org/apache/druid/query/policy/RestrictAllTablesPolicyEnforcer.java:
```diff
@@ -0,0 +1,109 @@
+    if (allowedPolicies == null) {
+      this.allowedPolicies = ImmutableList.of();
+    } else {
+      Set<String> policyClazz = new Reflections("org.apache.druid.query.policy").getSubTypesOf(
```
I'm not totally sure this will work for extensions. At any rate, I think it isn't really necessary, since its purpose appears to be just to verify that `allowedPolicies` only contains valid classes. It's ok if it contains invalid classes.

cecemei replied:
yea extensions would need to make code changes here, or maybe we need to ask for the full path.

cecemei replied:
updated this to use a ClassLoader to load from full path, this should be more helpful.

gianm commented on processing/src/main/java/org/apache/druid/query/policy/RestrictAllTablesPolicyEnforcer.java:
```diff
@@ -0,0 +1,109 @@
+import java.util.stream.Collectors;
+
+/**
+ * All tables must be restricted by a policy. 
```
The enforcer is actually more strict than this, I think. It appears to forbid any compound datasources, so for example, it wouldn't allow `unnest` -> `restricted` -> `table`. That's fine, since it's just here for enabling unit tests. But this should be called out in the javadocs and should be reflected in the name of the class. IMO, `RestrictedTablesOnlyEnforcer` is more clear than `RestrictAllTablesPolicyEnforcer`.

cecemei replied:
unnest union are handled now, since we're calling datasource.validate(enforcer)

gianm commented on processing/src/main/java/org/apache/druid/segment/RestrictedSegment.java:
```diff
@@ -42,13 +43,11 @@
-  protected final SegmentReference delegate;
-  protected final Policy policy;
+  @VisibleForTesting
+  public final SegmentReference delegate;
```
I don't see where this is being used in tests. Is it meant to be?

cecemei replied:
true, reverted this

gianm commented on processing/src/test/java/org/apache/druid/query/policy/RestrictAllTablesPolicyEnforcerTest.java:
```diff
@@ -0,0 +1,79 @@
+    RestrictedDataSource restricted = RestrictedDataSource.create(table, policy);
+
+    Assert.assertFalse(policyEnforcer.validate(table));
+    Assert.assertTrue(policyEnforcer.validate(restricted));
```
Please include tests for other kinds of datasources, such as `query`, `join`, `unnest`, `inline`, etc.

cecemei replied:
changed walker to call ds.validate(enforcer), and it validates all children, so enforcer only deal with TableDS and RestrictedDS. there're more tests in `DataSourceTest`, also going to add one for `inline` soon

gianm replied:
ok, TY

gianm commented on processing/src/main/java/org/apache/druid/query/policy/PolicyEnforcer.java:
```diff
@@ -0,0 +1,45 @@
+    @JsonSubTypes.Type(value = RestrictAllTablesPolicyEnforcer.class, name = "restrictAllTables"),
+})
+@UnstableApi
+public interface PolicyEnforcer
```
IMO it should also be possible for an enforcer to validate the `Segment` that is about to be processed by the query stack. This provides an additional layer of sanity checking after the `DataSource` has been used to decorate the physical `Segment` and produce a mapped `Segment`.

Suggested method:

```java
boolean validate(Segment segment);
```

It should be called:

- for native queries, right before passing a `Segment` to `QueryRunnerFactory#createRunner`.
- for MSQ, in `BaseLeafFrameProcessor#runIncrementally` before calling `runWithSegment`.

cecemei replied:
validate(Segment, Policy) has been added, unit tests are look good

gianm reviewed (commented):
In addition to the line comments, please include MSQ validation of `DataSource` as well (& tests for the same). Like the validations for native queries, it should be called when the query is finished planning and is about ready to be executed.

gianm commented on processing/src/main/java/org/apache/druid/query/policy/PolicyEnforcer.java:
```diff
@@ -0,0 +1,45 @@
+import org.apache.druid.query.DataSource;
+
+/**
+ * A mechanism for enforcing policies on data sources in Druid queries.
```
One more comment: there should be something here that notes that the `PolicyEnforcer` is really meant as a sanity checker, and not a primary authorization mechanism. It should not be used to implement security rules, only verify that they have been implemented correctly as a last line of defense against incorrect policy usage.

gianm commented on server/src/main/java/org/apache/druid/client/CachingClusteredClient.java:
```diff
@@ -199,6 +205,11 @@ private <T> Sequence<T> run(
   {
+    Query<T> query = queryPlus.getQuery();
+    if (!(query instanceof SegmentMetadataQuery) && !policyEnforcer.validate(query.getDataSource())) {
+      throw new ISE("Failed security validation with dataSource [%s]", query.getDataSource());
```
Let's make this a `DruidException` with category `FORBIDDEN` and persona `OPERATOR`.

cecemei replied:
updated

gianm commented on server/src/main/java/org/apache/druid/server/coordination/ServerManager.java:
```diff
@@ -190,6 +195,10 @@ public <T> QueryRunner<T> getQueryRunnerForSegments(Query<T> theQuery, Iterable<
     }
 
+    if (!(theQuery instanceof SegmentMetadataQuery) && !policyEnforcer.validate(dataSourceFromQuery)) {
+      throw new ISE("Failed security validation with dataSource [%s]", dataSourceFromQuery);
```
Let's make this a `DruidException` with category `FORBIDDEN` and persona `OPERATOR`.

cecemei replied:
updated

cecemei commented on server/src/main/java/org/apache/druid/server/coordination/ServerManager.java:
```diff
@@ -190,6 +195,10 @@ public <T> QueryRunner<T> getQueryRunnerForSegments(Query<T> theQuery, Iterable<
       throw new ISE("Cannot handle subquery: %s", dataSourceFromQuery);
     }
 
+    if (!(theQuery instanceof SegmentMetadataQuery) && !policyEnforcer.validate(dataSourceFromQuery)) {
```
I think this is from when policy is first rolled out we're not sure what to do with `SegmentMetadataQuery`, nevertheless `NoRestrictionPolicy` should be allowed since it's coming from internal, PR to allow it: https://github.com/apache/druid/pull/17864

also updated this pr to remove this special case

gianm commented on integration-tests/docker/environment-configs/common:
```diff
@@ -39,6 +39,7 @@ druid_auth_authenticatorChain=["basic"]
 druid_auth_authorizer_basic_type=basic
 druid_auth_authorizers=["basic"]
 druid_auth_authorizeQueryContextParams=true
+druid_policy_enforcer_type=none
```
Let's not include this here, same reason as we shouldn't include in the default configs.

gianm commented on processing/src/test/java/org/apache/druid/query/DataSourceTest.java:
```diff
@@ -221,4 +225,53 @@ public void testMapWithRestriction_onRestrictedDataSource_alwaysThrows()
+  }
+
+  @Test
+  public void testValidate_onUnionDataSource()
```
Please include a version of this where there is a `union` and both children are `restricted`. In this case would `RestrictAllTablesPolicyEnforcer` be able to pass?

Please also include one with a `join`.

cecemei replied:
Added tests for union with restricted and join.

gianm commented on server/src/main/java/org/apache/druid/guice/security/PolicyModule.java:
```diff
@@ -0,0 +1,37 @@
+  @Override
+  public void configure(Binder binder)
+  {
+    JsonConfigProvider.bind(binder, "druid.policy.enforcer", PolicyEnforcer.class);
```
Ah, maybe you had to include the config in the default files because there is no default here. This is the right place to put a default, you can do it with `JsonConfigProvider.bindWithDefault`. It doesn't have the same problem as Jackson's `defaultImpl`: with the default here, if the type is misspelled then the user will get an error rather than silently using the default impl.

cecemei replied:
update to use bindWithDefault, also updated `JsonConfigurator` to be able to construct an instance from `JsonCreator` annotated factory method

gianm commented on processing/src/main/java/org/apache/druid/query/policy/PolicyEnforcer.java:
```diff
@@ -0,0 +1,67 @@
+
+  /**
+   * Validate a {@link DataSource} node against the policy enforcer. This method is called during query planning from
+   * {@link DataSource#validate(PolicyEnforcer)}.
```
The javadoc here should specify how it's called and what the `ds` argument will be. I think with the new structure, where it is called indirectly through `DataSource` itself, that means this `validate` method will get called once for each `restricted` and once for each `table` that is not wrapped under a `restricted`.

cecemei replied:
added in javadoc

gianm commented on processing/src/main/java/org/apache/druid/query/policy/PolicyEnforcer.java:
```diff
@@ -0,0 +1,67 @@
+
+
+  /**
+   * Validate a {@link Segment} node against the policy enforcer. This method is called during query planning from
```
Similar thought here, the javadoc should specify how it's called and what the `segment` argument will be.

cecemei replied:
updated javadoc

gianm commented on processing/src/main/java/org/apache/druid/query/policy/PolicyEnforcer.java:
```diff
@@ -0,0 +1,67 @@
+   * @param ds data source to validate
+   * @return true if the data source is valid according to the policy enforcer, false otherwise
+   */
+  boolean validate(DataSource ds, Policy policy);
```
I see a few call sites missing for this:

- it isn't being called on realtime tasks (equivalent of `ServerManager` there is `SinkQuerySegmentWalker`)
- it isn't being called for any MSQ paths-- for this I think it would work to call it in `BaseLeafFrameProcessorFactory#makeProcessors` right before creating the `segmentMapFnProcessor`. Actually, you could also move `validate(Segment)` into the map fn itself, which would simplify enforcement. All the enforcement logic could be in the `BaseLeafFrameProcessorFactory` class.

Please add tests that cover these.

cecemei replied:
added new tests for msq

gianm commented on processing/src/main/java/org/apache/druid/query/policy/PolicyEnforcer.java:
```diff
@@ -0,0 +1,67 @@
+   * @param segment segment to validate
+   * @return true if the segment is valid according to the policy enforcer, false otherwise
+   */
+  boolean validate(Segment segment, Policy policy);
```
I believe a call site for this is missing on realtime tasks. There should be something in `SinkQuerySegmentWalker`. Please add a test that covers that too.

cecemei replied:
added

gianm reviewed (commented):
TY this is looking good so far. I think the change to make the enforcer only have to deal with Restricted and Table makes sense.

cecemei commented:
> TY this is looking good so far. I think the change to make the enforcer only have to deal with Restricted and Table makes sense.

Addressed all your comments, and last round of unit/integration test looks good, could you please take a look again? Thx.

gianm commented on processing/src/main/java/org/apache/druid/segment/SegmentReference.java:
```diff
@@ -27,4 +29,15 @@
+   * <p>
+   * This should be called right before the segment is about to be processed by the query stack.
+   */
+  default boolean validate(PolicyEnforcer policyEnforcer)
```
A call is missing on the realtime path; there's one in `ReferenceCountingSegmentQueryRunner` but that's only used by Historicals.

Similar to the idea about tying together `DataSource.validate` with `withPolicies`, we could simplify this by tying it together with the application of functions created by `createSegmentMapFunction`.

i.e., create a utility method that applies a segment map function `Function<SegmentReference, SegmentReference>` and also calls `validate` on the result, then use that utility method everywhere we apply segment map functions.

or this might be easier to integrate in the current code: create a utility method that decorates a segment map function with a validator. i.e. it takes `Function<SegmentReference, SegmentReference>` + `PolicyEnforcer`, and it returns `Function<SegmentReference, SegmentReference>` which embeds the validation call.

cecemei replied:
Updated this:
- On the runner/processor, enforcer would check on the mapped segment, this is in `ExecutionVertex.createSegmentMapFunction`
- Prior to execution on date node, the enforcer works in `QueryLifecycle` and `DruidQuery`, through `withPolicies`.

The exception is also centralized in `PolicyEnforcer`.

cecemei commented on server/src/main/java/org/apache/druid/client/CachingClusteredClient.java:
```diff
@@ -199,6 +204,13 @@ private <T> Sequence<T> run(
   )
   {
+    Query<T> query = queryPlus.getQuery();
+    if (!query.getDataSource().validate(policyEnforcer)) {
```
Updated the enforcer to be called within `withPolicies`

cecemei commented on server/src/main/java/org/apache/druid/server/coordination/ServerManager.java:
```diff
@@ -189,6 +194,12 @@ public <T> QueryRunner<T> getQueryRunnerForSegments(Query<T> theQuery, Iterable<
       throw new ISE("Cannot handle subquery: %s", dataSourceFromQuery);
     }
 
+    if (!dataSourceFromQuery.validate(policyEnforcer)) {
```
all exception is in `PolicyEnforcer` now

cecemei commented on processing/src/main/java/org/apache/druid/query/policy/RestrictAllTablesPolicyEnforcer.java:
```diff
@@ -0,0 +1,136 @@
+                                                        }
+                                                      })
+                                                      .collect(Collectors.toList());
+      Preconditions.checkArgument(
```
removed the reflection and class loader here, now `allowedPolicies` should just be the full path to policy class

cecemei commented on processing/src/main/java/org/apache/druid/query/policy/PolicyEnforcer.java:
```diff
@@ -0,0 +1,70 @@
+   * @param policy  the policy on the segment, {@link org.apache.druid.segment.RestrictedSegment#policy} or null for other
+   * @return {@code true} if the segment complies with the policy, {@code false} otherwise
+   */
+  boolean validate(SegmentReference segment, Policy policy);
```
the `PolicyEnforcer` class has been updated, i think it's more obvious where the validateOrElseThrow is called from.

cecemei commented on processing/src/main/java/org/apache/druid/query/policy/PolicyEnforcer.java:
```diff
@@ -0,0 +1,70 @@
+{
+
+  /**
+   * Validates a {@link DataSource} against the policy enforcer. This method is invoked before query execution
```
Updated the enforcer to take `TableDataSource` and `ReferenceCountingSegment` directly, it should be more obvious how the `validateOrElseThrow` is called.

cecemei commented:
> > TY this is looking good so far. I think the change to make the enforcer only have to deal with Restricted and Table makes sense.
> 
> Addressed all your comments, and last round of unit/integration test looks good, could you please take a look again? Thx.

Hey, this is ready for review again, a few things to look for from last round:
- PolicyEnforcer class now take TableDataSource and ReferenceCountingSegment directly, this should make call stack more obvious. The exception is also thrown from enforcer. 
- The enforcer is injected into `QueryLifecycleFactory` and `TaskToolboxFactory`, and used in `withPolicies` and  `createSegmentMapFunction`. Some file changes are purely because of the constructor now asks for enforcer.

gianm commented on processing/src/main/java/org/apache/druid/segment/SegmentReference.java:
```diff
@@ -19,12 +19,31 @@
+   * This should be called right before the segment is about to be processed by the query stack, and after
+   * {@link org.apache.druid.query.planning.ExecutionVertex#createSegmentMapFunction(PolicyEnforcer)}.
+   */
+  default void validateOrElseThrow(PolicyEnforcer policyEnforcer)
```
There is a call missing at `BroadcastJoinSegmentMapFnProcessor` -- the path MSQ goes down when there is a segment-level join happening. Maybe move the `PolicyEnforcer` arg from `ExecutionVertex` to `DataSource#createSegmentMapFunction` itself?

Or perhaps have `BroadcastJoinSegmentMapFnProcessor` apply the enforcer on its own. If you go that route, please also add a comment to the javadoc for `DataSource#createSegmentMapFunction` reminding callers that if they are using this method instead of `ExecutionVertex`, they need to apply a `PolicyEnforcer` to the resulting segment.

cecemei replied:
Looking at `BroadcastJoinSegmentMapFnProcessor`, it converts `InputNumberDataSource` to `InlineDataSource`, but the enforcer won't have effects on inline data.

cecemei replied:
I managed to run some tests based on BroadcastJoin in MSQ, disabled the enforcer on `DruidQuery` side, and it's actually catching security validation failure when I use `TableDataSource` as one of left child and right child, in the scanning step. And the join step (runWithInputChannel) actually don't enforce anything, because it's using a `FrameSegment`.

gianm replied:
> Looking at `BroadcastJoinSegmentMapFnProcessor`, it converts `InputNumberDataSource` to `InlineDataSource`, but the enforcer won't have effects on inline data.

It only inlines if the input number is in the `inputNumberToProcessorChannelMap` map. The base input won't be in that map, and that base input might be a regular table. I think in the simplest case where a regular table is joined with some subquery result, like `select page, q.c from wikipedia, (select count(*) c from "kttm-v2-2019-08-25") q`, the segments from `wikipedia` won't be checked since they'll be using the map function generated here.

I tried setting a breakpoint at `validateOrElseThrow(ReferenceCountingSegment segment, Policy policy)` and did find that. For MSQ tasks, the breakpoint triggered for `kttm-v2-2019-08-25` but not for `wikipedia`.

For Dart the situation was a little worse; the breakpoint for the `validateOrElseThrow(ReferenceCountingSegment segment, Policy policy)` method triggered also only for `kttm-v2-2019-08-25`, but then `validate` wasn't called because Dart has two layers of `ReferenceCountingSegment`, so the check `baseSegment instanceof QueryableIndexSegment || baseSegment instanceof IncrementalIndexSegment` is `false`.

IMO, the best approach to the first issue (segment map fn created by `BroadcastJoinSegmentMapFnProcessor` not having enforcement) is to move the `PolicyEnforcer` into `DataSource#createSegmentMapFunction`, just like `withPolicies`. That way we never need to think about whether a particular bare call to `DataSource#createSegmentMapFunction` is safe or not, because all calls would need the enforcer present.

For the second issue (Dart not validating properly because of double-wrapped segments) -- perhaps it would make sense to move `validateOrElseThrow` from `ReferenceCountingSegment` to `Segment`, and have the impl in `ReferenceCountingSegment` call `getBaseSegment().validateOrElseThrow(policyEnforcer)` rather than `validateOrElseThrow(this, policyEnforcer)`?

cecemei replied:
Thanks for the example, I was able to reproduce the error now, it looks like right child is inlined but it has already gone through a scan stage, and the current implementation missed the check on the left child. 

I'm think, we can ask `BroadcastJoinSegmentMapFnProcessor` to create an `ExecutionVertex` based on the query, and pass enforcer in `ExecutionVertex`'s `createSegmentMapFunction`, so like 
```
DataSource transformed = inlineChannelData(query.getDataSource());
ExecutionVertex ev = ExecutionVertex.of(query.withDataSource(transformed));
return ev.createSegmentMapFunction(policyEnforcer);
```
This way we can centralize the enforcer in `ExecutionVertex` without changing `DataSource` interface. 

For the Dart issue, I didn't know it can wrap two layer of `ReferenceCountingSegment`, I think maybe we can add recursive in the validation? For a restricted segment, would it be 

- Restrict
  - Ref
     - Ref_base
  -  policy

or 

- Ref
  - Restrict 
     - Ref_base
     - policy

?

cecemei replied:
I guess the first case is easy we can get keep use `getBase`, but the second is tricky since restrict doesn't expose the delegate.

cecemei replied:
Is it possible to be:
- Restrict
  - Restrict
     - Ref_base
     - policy
  - policy

cecemei replied:
Updated the PR to handle `BroadcastJoinSegmentMapFnProcessor` and multiple layer of wrapping in Dart, so case 1 and case 2 can both pass validation, and case 3 would fail.

gianm replied:
The changes to `BroadcastJoinSegmentMapFnProcessor` look good to me.

The implementation and javadoc of `PolicyEnforcer#validateOrElseThrow` for `ReferenceCountingSegment` seems a little sketchy still, for a couple reasons:

- The "else skip validation" seems like a hole for stuff to fall into. Prior to the latest fix, an inner `ReferenceCountingSegment` did fall into it. I'm wondering if other things could potentially fall in, like extension segment types, new segment types, etc.
- The branch doing `instanceof QueryableIndexSegment || instanceof IncrementalIndexSegment` isn't strong enough to detect whether a `Segment` corresponds to a regular table, because extensions can add other kinds of segments backing tables. This is a specific kind of thing that could fall into the hole mentioned previously.
- I don't understand why the `instanceof RestrictedSegment` branch is needed -- is it really possible for RestrictedSegment to wrap another RestrictedSegment? I would think this is impossible given that RestrictedDataSource can only wrap a TableDataSource.

I observe that the `DataSource` version of validation is cleaner and more robust, and can be because there are two important things `DataSource` has:

- `DataSource#getChildren` exists, so datasource trees can be walked robustly
- We know that tables correspond to `TableDataSource`, so it's always possible to identify whether a leaf datasource is a regular table or not.

To have equally robust validation for `Segment`, I think we would need to add `Segment#getChildren` to address the first point. For the second point I think we can use a check like this for leaf segments to see if they correspond to regular tables: `as(PhysicalSegmentInspector.class) != null`.

I tried to think for a while of a robust way to write this code without adding `Segment#getChildren` and wasn't able to. So I do think we should do that. The default implementation should throw an unsupported operation error.

cecemei replied:
I think `SegmentReference` could serve the purpose of `Segment#getChildren`. The default impl is to throw, and every extension (`ReferenceCountingSegment`, `HashJoinSegment`, `WrappedSegmentReference` needs to implement its own validation (basically asking delegate to validate).

To solidify, I'm thinking, can we eliminate the use cases when `ReferenceCountingSegment` is wrapped with another `SegmentReference`? I think it would be simpler if we know the structure is `RestrictedSegment` wraps a `ReferenceCountingSegment`, which wraps a basic `segment` (`QueryableIndexSegment`, `LookupSegment`, etc). 

It seems tests using `SpecificSegmentsQuerySegmentWalker` follows this structure as well.

gianm replied:
> To solidify, I'm thinking, can we eliminate the use cases when `ReferenceCountingSegment` is wrapped with another `SegmentReference`? I think it would be simpler if we know the structure is `RestrictedSegment` wraps a `ReferenceCountingSegment`, which wraps a basic `segment` (`QueryableIndexSegment`, `LookupSegment`, etc).

I did see double-wrapping (`ReferenceCountingSegment` on top of `ReferenceCountingSegment`) with Dart with the query `select page, q.c from wikipedia, (select count(*) c from "kttm-v2-2019-08-25") q`. However, I doubt this is "necessary". The Dart code could likely be adjusted to not do this. So, if it's helpful, go ahead and restrict things so `ReferenceCountingSegment` cannot wrap another `SegmentReference`.

> To have equally robust validation for `Segment`, I think we would need to add `Segment#getChildren` to address the first point. For the second point I think we can use a check like this for leaf segments to see if they correspond to regular tables: `as(PhysicalSegmentInspector.class) != null`.

About this (using `PhysicalSegmentInspector` as a proxy for "Segment represents a regular table"), I thought about it some more and IMO this approach is still not ideal. Nothing in the interface says that it can't be implemented for a non-table, and nothing requires that it is implemented for a table.

I think a nicer idea would be to make `Segment#getId` nullable, and spec it so that it should return nonnull for regular tables backed by actual segments, null for anything else. This to me would make more sense than the current way `getId()` works, because currently dummy IDs are needed for `Segment` that aren't backed by actual segments, which seems odd. I skimmed usages of `getId()`; most seem to either be running in scenarios where it would always be nonnull anyway, or else are using it to interpolate into log messages (could use `toString` or `asString` instead).

gianm replied:
In the interests of having this PR merged before it grows too large, I suggest doing a super-strict check now -- validate all leaf segments regardless of whether they are actual tables or not. This will essentially mean that validation will always fail on lookups, external, and inline segments, because they will not have policies applied. In a follow up we can add something to constraint validation to regular-table-backed `Segment` only, perhaps using this nullable `Segment#getId` idea.

cecemei replied:
Completely agree on the `SegmentId` approach, would be a cleaner solution and we can have a better mapping on different datasource (table, inline, lookup, external) to segmentId. 

I reverted the double-wrapping check in commit 1e6632fcb779919f3fc54050ab9b44aba19d3265, and then actually moved the no-double-wrapping check f604abffa9f34a7b7c520445a1cce14d92325088 to another pr (https://github.com/apache/druid/pull/17943). Please take a look again!

gianm reviewed (commented):
The new structure of `PolicyEnforcer` is imo a lot clearer.

gianm reviewed (approved):
LGTM. Items left to follow-ups, as I recall, are:

- Some way to identify if a Segment corresponds to a regular table (possibly by leveraging `getId`)
- Consider restricting double-wrapping of segment references (#17943)
